File: workflow-sdk-node/src/client/rest.ts
import { fetch } from 'undici'
import type { EngineClient, Job, PollJobsRequest } from './types.js'

/**
 * REST transport implementation of {@link EngineClient}.
 * Uses `undici` fetch for HTTP/1.1 requests.
 */
export class RestEngineClient implements EngineClient {
  private readonly baseUrl: string

  constructor(baseUrl: string) {
    this.baseUrl = baseUrl.replace(/\/$/, '')
  }

  async pollJobs(req: PollJobsRequest): Promise<Job[]> {
    const resp = await fetch(`${this.baseUrl}/v1/jobs/poll`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        workerId: req.workerId,
        jobTypes: req.jobTypes,
        maxJobs: req.maxJobs,
      }),
    })
    if (!resp.ok) {
      throw new Error(`pollJobs: engine returned ${resp.status}`)
    }
    const data = (await resp.json()) as { jobs?: Job[] }
    return data.jobs ?? []
  }

  async completeJob(
    jobId: string,
    workerId: string,
    variables?: Record<string, unknown>,
  ): Promise<void> {
    const resp = await fetch(`${this.baseUrl}/v1/jobs/${jobId}/complete`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ workerId, variables }),
    })
    if (!resp.ok && resp.status !== 204) {
      throw new Error(`completeJob ${jobId}: engine returned ${resp.status}`)
    }
  }

  async failJob(
    jobId: string,
    workerId: string,
    errorMessage: string,
    retryable: boolean,
  ): Promise<void> {
    const resp = await fetch(`${this.baseUrl}/v1/jobs/${jobId}/fail`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ workerId, errorMessage, retryable }),
    })
    if (!resp.ok && resp.status !== 204) {
      throw new Error(`failJob ${jobId}: engine returned ${resp.status}`)
    }
  }
}

File: workflow-sdk-go/client/rest.go
package client

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

// RestClient implements EngineClient via HTTP/REST.
type RestClient struct {
	baseURL    string
	workerID   string
	httpClient *http.Client
}

// NewRest creates a RestClient targeting baseURL.
func NewRest(baseURL, workerID string) *RestClient {
	return &RestClient{
		baseURL:    baseURL,
		workerID:   workerID,
		httpClient: &http.Client{},
	}
}

// PollJobs implements EngineClient.
func (c *RestClient) PollJobs(ctx context.Context, req PollJobsRequest) ([]Job, error) {
	body, _ := json.Marshal(map[string]any{
		"workerId": req.WorkerID,
		"jobTypes": req.JobTypes,
		"maxJobs":  req.MaxJobs,
	})
	resp, err := c.post(ctx, "/v1/jobs/poll", body)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("poll: engine returned %d", resp.StatusCode)
	}
	var envelope struct {
		Jobs []Job `json:"jobs"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&envelope); err != nil {
		return nil, fmt.Errorf("poll: decode response: %w", err)
	}
	return envelope.Jobs, nil
}

// CompleteJob implements EngineClient.
func (c *RestClient) CompleteJob(ctx context.Context, jobID, workerID string, variablesToSet map[string]any) error {
	body, _ := json.Marshal(map[string]any{
		"workerId":  workerID,
		"variables": variablesToSet,
	})
	resp, err := c.post(ctx, "/v1/jobs/"+jobID+"/complete", body)
	if err != nil {
		return err
	}
	resp.Body.Close()
	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
		return fmt.Errorf("completeJob %q: engine returned %d", jobID, resp.StatusCode)
	}
	return nil
}

// FailJob implements EngineClient.
func (c *RestClient) FailJob(ctx context.Context, jobID, workerID, errorMessage string, retryable bool) error {
	body, _ := json.Marshal(map[string]any{
		"workerId":     workerID,
		"errorMessage": errorMessage,
		"retryable":    retryable,
	})
	resp, err := c.post(ctx, "/v1/jobs/"+jobID+"/fail", body)
	if err != nil {
		return err
	}
	resp.Body.Close()
	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
		return fmt.Errorf("failJob %q: engine returned %d", jobID, resp.StatusCode)
	}
	return nil
}

func (c *RestClient) post(ctx context.Context, path string, body []byte) (*http.Response, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.baseURL+path, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	return c.httpClient.Do(req)
}
